### PR TITLE
Add Persian Calendar (Solar Hijri)

### PR DESCRIPTION
**Pull request by asadoll**
This is a first attempt in order to implement **Persian Calendar** complying with the **[date](https://github.com/HowardHinnant/date)** library.
The "solar_hijri.h" is structured almost like "islamic.h"; There are no real tests written atm but there are a few logging and review helpers in test/solar_hijri_test/parse.pass.cpp

### CODE REVIEW

**Comment by HowardHinnant** [comment]
Thanks for doing this work!

**Comment by asadoll** [comment]
Thanks for all the work you and others have done.

**Comment by asadoll** [comment]
solar_hijri::year_month_day day point conversion test is failing before UTC, both in my local setup (UTC +03:30) and CI.

**Comment by asadoll** [comment]
@HowardHinnant, Persian Calendar tests are passing but it seems on [this build environment](https://travis-ci.org/HowardHinnant/date/jobs/647464019#L752) the tests are not finding the time-zone database. Any advice?

**Comment by HowardHinnant** [comment]
My first guess was that you didn't have this recent commit: 9a0ee2542848ab8625984fc8cdbfb9b5414c0082

But I see that you do.  So don't worry about the travis failure.  That won't block this PR.

**Comment by asadoll** [comment]
Any urgent/special tests are there to write? I've refactored arithmetic and day point conversion test so far.

**Comment by HowardHinnant** [comment]
Sorry I haven't reviewed this yet.  It's just lack of free time on my part.

I was going to suggest a daily sweep test over a ridiculous range testing conversions both to and from sys_days.  But it looks like you've already done that in year_month_day.pss.cpp.  This looks like really nice work!

**Comment by HowardHinnant** [comment]
I think this would have to be a separate PR, but how about an html file targeted to the gh-pages branch that documents your calendar.  Doesn't have to be extensive.  Maybe along the lines of https://github.com/HowardHinnant/date/blob/gh-pages/islamic.html .  Just a few examples of creation and interoperability.  Such a file shows up in readable format at a url such as this:  https://howardhinnant.github.io/date/islamic.html

**Comment by HowardHinnant** [comment]
```
/Users/howardhinnant/Development/date/include/date/solar_hijri.h:2213:16: warning: unused variable 'years_in_first_cycle' [-Wunused-variable]
    auto const years_in_first_cycle = static_cast<unsigned>(29);
```

**Comment by HowardHinnant** [comment]
```
year_month_day.pass.cpp:157:26: warning: unused variable 'dp2' [-Wunused-variable]
                sys_days dp2 = ymd2;
                         ^
```

**Comment by HowardHinnant** [comment]
In parse.pass.cpp, need to convert all of the output to `cout` into tests for the correct values.

**Comment by HowardHinnant** [comment]
Except for adding the header solar_hijri.h, this PR should not modify CMakeLists.txt.

File: .gitignore [hunk 1]
```diff
@@ -192,3 +192,4 @@ nbproject/*
 *.cd
 a.out
 cmake-build-debug/*
+.ccls-cache/*
```
Comment (HowardHinnant) [inline .gitignore]: What does this do, and can we do without it?  I'd rather get a clean "add calendar" build, and leave everything else (such as build changes) to a separate PR.

File: include/date/solar_hijri.h [hunk 2]
```diff
@@ -2198,16 +2202,46 @@ year_month_day::to_days() const NOEXCEPT
...
+    // auto const era = static_cast<int>(era_d);
+    // // auto const era = static_cast<int>(((y-475) >= 0 ? (y-475) : (y-475)-(years_in_era-1)) / years_in_era);
+    // auto const fdoe = static_cast<int>(epoch + era * days_in_era);
+    // auto const yoe = static_cast<unsigned>(y - era * years_in_era);
+    // auto const fdoyoe = static_cast<double>(yoe*days_in_year);
+    // // auto const fdoyoe = static_cast<unsigned>(yoe*days_in_year);
+    // auto const fdoyoe_i = static_cast<int>(fdoyoe);
+    // auto const doy = 30*(m-1) + ((m > 6) ? 6 : m-1) + d-1;                     // [0, 365]
+    // auto const doe = fdoe + fdoyoe_i + doy;
+    // return days{doe - unix_time_shift};
```
Comment (HowardHinnant) [inline include/date/solar_hijri.h]: Once you've achieved confidence in this new algorithm, can you remove the old one before we commit to master?

File: include/date/solar_hijri.h [hunk 3]
```diff
@@ -39,7 +42,7 @@ static const auto days_in_era = static_cast<unsigned>(1029983);
 static const auto years_in_era = static_cast<unsigned>(2820);
 static constexpr auto days_in_year = 365.24219858156028368;
 static constexpr auto leap_threshold = 0.24219858156028368;
```
Comment (HowardHinnant) [inline include/date/solar_hijri.h]: `leap_threshold` looks unused.

File: include/date/solar_hijri.h [hunk 4]
```diff
@@ -39,7 +42,7 @@ static const auto days_in_era = static_cast<unsigned>(1029983);
...
-static const auto unix_time_shift = static_cast<unsigned>(2440587);
+static const auto unix_time_shift = static_cast<unsigned>(2440588);
```
Comment (HowardHinnant) [inline include/date/solar_hijri.h]: Are these constants part of the interface or just implementation details?  If the latter, how about stuffing them into a sub-namespace?

**Comment by HowardHinnant** [comment]
Can we remove `year_month_day::days_in_month`?

**Comment by HowardHinnant** [comment]
Just fyi, the committee decided they didn't like "mon" and "jan" and asked me to rename them to "Monday" and "January".  I left in "mon" and "jan" strictly for backwards compatibility for my clients.  This is just informative.  `solar_hijri` is your calendar, and you can make the names whatever style you want, and have as many different aliases for the names as you want.

File: .gitignore (hunk 1 above)
Comment (asadoll) [reply .gitignore]: It's the cache folder from [ccls](https://github.com/MaskRay/ccls) which I'm using in Emacs.

File: include/date/solar_hijri.h (hunk 2 above)
Comment (asadoll) [reply include/date/solar_hijri.h]: Done!

File: include/date/solar_hijri.h (hunk 4 above)
Comment (asadoll) [reply include/date/solar_hijri.h]: Done!

File: .gitignore (hunk 1 above)
Comment (HowardHinnant) [reply .gitignore]: Any harm if I remove it prior to pushing to master?

Comment (asadoll) [reply .gitignore]: Not at all.

**Comment by HowardHinnant** [comment]
Merged: https://github.com/HowardHinnant/date/commit/9cc3a7bca17ba2ef9041255d6ccab89b085fc9e4

**Comment by HowardHinnant** [comment]
Would still love to see docs on the gh-pages branch.

**Comment by asadoll** [comment]
Was busy with some other projects, will make a PR very soon.

**Pull request closed without merging**